### PR TITLE
[Doc] Update documentation

### PR DESCRIPTION
**Pull request by neozhaoliang**
Update docs for ndarray, kernels and dataclass

### CODE REVIEW

File: docs/lang/articles/advanced/dataclass.md [hunk 1]
```diff
@@ -12,6 +12,12 @@ Taichi provides custom [struct types](../type_system/type.md#compound-types) for
 
 To achieve the ends, Taichi enabled the `@ti.dataclass` decorator on a Python class. This is inspired by Python's [dataclass](https://docs.python.org/3/library/dataclasses.html) feature, which uses class fields with annotations to create data types.
 
+:::note
+
+The `dataclass` in Taichi is simply a wrapper for `ti.types.struct`. Therefore, the member types that a `dataclass` object can contain are the same as those allowed in a struct. They must be one of the following types: scalars, matrix/vector types, and other dataclass/struct types. Objects like `field`, `Vector field`, and `Ndarray` cannot be used as members of a `dataclass` object.
```
Comment (jim19930609) [inline docs/lang/articles/advanced/dataclass.md:17]: Correct me if I were wrong, but I thought `field` and `Ndarray` are not allowed in ti.struct? @lin-hitonami

File: docs/lang/articles/kernels/kernel_function.md [hunk 2]
```diff
@@ -348,3 +348,23 @@ Type hinting is a formal solution to statically indicate the type of value withi
...
+    print(ti.sin(1.0))
+
+test()
+
+ti.init(arch=ti.gpu)
+
+test()
+```
+
+In the above code, we initially specify the CPU as the backend and run the `test` function. At this point, the `test` function runs on the CPU backend. However, when we continue by calling `ti.init(arch=ti.gpu)` and specify the GPU as the backend, subsequent calls to `test` will result in the recompilation of the `test` kernel for the GPU backend and its subsequent execution on the GPU. In summary, Taichi does not support running multiple kernels on different backends simultaneously.
```
Comment (jim19930609) [inline docs/lang/articles/kernels/kernel_function.md:370]:
```suggestion
In the provided code, we begin by designating the CPU as the backend, upon which the test function operates. Notably, the test function is initially executed on the CPU backend. As we proceed by invoking ti.init(arch=ti.gpu) to designate the GPU as the backend, all ensuing invocations of test trigger a recompilation of the test kernel tailored for the GPU backend, subsequently executing on the GPU. To conclude, Taichi does not facilitate the concurrent operation of multiple kernels on varied backends.
```

File: docs/lang/articles/kernels/kernel_sync.md [hunk 3]
```diff
@@ -53,8 +53,10 @@ print(end - start)
 Most of the time, Taichi automatically handles data synchronization. For example, Taichi will automatically call `ti.sync()` to synchronize data in the following cases:
 
 1. The kernel has a return value.
-2. The `x.to_numpy()` method is called in the Python scope, and `x` is being used by other kernels.
-3. An attempt is made to access or modify a Taichi field in the Python scope, and this field is being used by other kernels.
+2. The `x.to_numpy()` method is called in the Python scope, and `x` is being used by other kernels. Taichi will wait for the kernel that is using `x` to finish before invoking the `x.to_numpy()` method.
+3. An attempt is made to access or modify a Taichi field in the Python scope, and this field is being used by other kernels. Taichi will wait for the kernel that is using `x` to finish before proceeding with any statements that access or modify `x`.
+
+In the example above, we need to explicitly call `ti.sync()` because its kernel does not have a return value, does not contain `print` statements, and does not involve any operations on fields.
```
Comment (jim19930609) [inline docs/lang/articles/kernels/kernel_sync.md:59]:
```suggestion
If the x.to_numpy() method is invoked in the Python scope while x is concurrently being utilized by other kernels, Taichi will ensure the completion of the kernel using x before the x.to_numpy() method is executed.

Similarly, in cases where there is an attempt to access or modify a Taichi field within the Python scope, and the field is concurrently being utilized by other kernels, Taichi will ensure the kernel that is using the field has finished prior to proceeding with any statements that access or modify the field.

In the provided example, it is necessary to explicitly call ti.sync(). This is due to the kernel lacking a return value, not containing any print statements, and not involving any operations on fields.
```

**Review by jim19930609 — APPROVED** [review]
+ Some minor polish via GPT-4

**Review by jim19930609 — COMMENTED** [review]
+ Some minor polish via GPT-4

File: docs/lang/articles/advanced/dataclass.md (hunk 1 above)
Comment (neozhaoliang) [reply docs/lang/articles/advanced/dataclass.md:17]: Is there a problem with this section, or we can resolve this comment?

Comment (lin-hitonami) [reply docs/lang/articles/advanced/dataclass.md:17]: They are not allowed in ti.struct and dataclass as described in this section.

**Pull request merged**